light/beacon/sync_committee_update.go

```diff
@@ -0,0 +1,544 @@
+	Committees [][]byte
+}
+
+// sctClient represents a peer that SyncCommitteeTracker sends signed heads and sync committee advertisements to
+type sctClient interface {
+	SendSignedHeads(heads []SignedHead)
+	SendUpdateInfo(updateInfo *UpdateInfo)
+}
+
+// sctServer represents a peer that SyncCommitteeTracker can request sync committee update proofs from
+type sctServer interface {
+	GetBestCommitteeProofs(ctx context.Context, req CommitteeRequest) (CommitteeReply, error)
+	CanRequest(updateCount, committeeCount int) bool
+	WrongReply(description string)
+}
+
+// UpdateInfo contains scores for an advertised update chain. Note that the most recent updates are always
+// advertised but earliest ones might not because of length limitation.
+type UpdateInfo struct {
+	AfterLastPeriod uint64
+	Scores          UpdateScores
+}
+
+// UpdateScore allows the comparison between updates at the same period in order to find the best
+// update chain that provides the strongest proof of being canonical.
+type UpdateScore struct {
+	signerCount, subPeriodIndex uint32
+	finalizedHeader             bool // update is considered finalized if has finalized header and 2/3 signatures
+}
+
+type UpdateScores []UpdateScore
+
+// isFinalized returns true if the update has a header signed by at least 2/3 of the committee,
+// referring to a finalized header that refers to the next sync committee. This condition is a
+// close approximation of the actual finality condition that can only be verified by full beacon nodes.
+func (u *UpdateScore) isFinalized() bool {
+	return u.finalizedHeader && u.signerCount > 341
+}
+
+// reorgRiskPenalty returns a modifier that approximates the risk of a non-finalized update header being reorged.
+// It should be subtracted from the signer participation count when comparing non-finaized updates.
+// This risk is assumed to be dropping exponentially as the update header is further away from the beginning of the period.
+func (u *UpdateScore) reorgRiskPenalty() uint32 {
+	return uint32(math.Pow(2, 10-float64(u.subPeriodIndex)/32))
+}
+
+// betterThan returns true if update u is considered better than w.
+func (u *UpdateScore) betterThan(w UpdateScore) bool {
+	uFinalized, wFinalized := u.isFinalized(), w.isFinalized()
+	if uFinalized || wFinalized {
+		// finalized update is always better than non-finalized; only signer count matters when both are finalized
+		if uFinalized && wFinalized {
+			return u.signerCount > w.signerCount
+		}
+		return uFinalized
+	}
+	// take reorg risk into account when comparing two non-finalized updates
+	return u.signerCount+w.reorgRiskPenalty() > w.signerCount+u.reorgRiskPenalty()
+}
+
+// Encode encodes the update score data in a compact form for in-protocol advertisement
+func (u *UpdateScore) Encode(data []byte) {
+	v := u.signerCount + u.subPeriodIndex<<10
+	if u.finalizedHeader {
+		v += 0x800000
+	}
+	var enc [4]byte
+	binary.LittleEndian.PutUint32(enc[:], v)
+	copy(data, enc[:3])
+}
```

Comment:
Where is the encoding format described?

Reply:
Now it's described in the `UpdateScore` description.